examples/migration/recursive_fibonacci/task_emulation_layer.h

```diff
@@ -55,49 +55,98 @@ class base_task {
     void operator() () const {
         base_task* parent_snapshot = m_parent;
         const_cast<base_task*>(this)->execute();
-        if (m_parent && parent_snapshot == m_parent) {
-            m_parent->release();
+        if (m_parent && parent_snapshot == m_parent && m_child_counter == 0) {
+            if (m_parent->remove_reference() == 0) {
+                m_parent->operator()();
+                delete m_parent;
+            }
+        }
+
+        if (m_child_counter == 0 && m_type == task_type::allocated) {
+            delete this;
         }
     }
 
     virtual void execute() = 0;
 
     template <typename C, typename... Args>
-    C* create_continuation(std::uint64_t ref, Args&&... args) {
+    C* allocate_continuation(std::uint64_t ref, Args&&... args) {
         C* continuation = new C{std::forward<Args>(args)...};
+        continuation->m_type = task_type::continuation;
         continuation->reset_parent(reset_parent());
         continuation->m_child_counter = ref;
         return continuation;
     }
 
     template <typename F, typename... Args>
     F create_child_of_continuation(Args&&... args) {
-        F obj{std::forward<Args>(args)...};
-        obj.reset_parent(this);
-        return obj;
+        return create_child_of_continuation_impl<F>(std::forward<Args>(args)...);
+    }
+
+    template <typename F, typename... Args>
+    F create_child_of_continuation_safe(Args&&... args) {
+        add_reference();
+        return create_child_of_continuation_impl<F>(std::forward<Args>(args)...);
+    }
+
+    template <typename F, typename... Args>
+    F* allocate_child_of_continuation(Args&&... args) {
+        return allocate_child_of_continuation_impl<F>(std::forward<Args>(args)...);
+    }
+
+    template <typename F, typename... Args>
+    F* allocate_child_of_continuation_safe(Args&&... args) {
```

Comment:
I think the "safe" part of the allocate_child_of_continuation_safe name is not correct since it solves a different problem than the recycle_as_safe_continuation in the old API. This function is more like allocate_additional_child_of in the old API. The "safe" in old API was about concurrent execution of about-to-be-recycled task with its children / predecessors.

Reply:
Do you have any suggestion for the names of these functions?

Reply:
I really don't like the old naming "allocate_additional_child_of" because that doesn't explain how it is different from "allocate_child_of", so let's not copy that.  And "safe" is not right either. Do we need the "of_continuation" on all of these calls; there's no other kind of child to allocate. What about "allocate_child_and_increment"? And call the other just "allocate_child".  Or how about "allocate_predecessor_and_increment" and "allocate_predecessor" -- perhaps these more clearly combines the concepts of child and continuation???